pkg/rancher/migrations.go

```diff
@@ -0,0 +1,76 @@
+package rancher
+
+import (
+	"github.com/mcuadros/go-version"
+	"github.com/rancher/rancher/pkg/auth/tokens"
+	v3 "github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"
+	rancherversion "github.com/rancher/rancher/pkg/version"
+	controllerv1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	k8serror "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+const (
+	cattleNamespace          = "cattle-system"
+	forceUpgradeLogoutConfig = "forceUpgradeLogout"
+	rancherVersionKey        = "rancherVersion"
+)
+
+// forceUpgradeLogout will delete all dashboard tokens forcing a logout.  This is useful when there is a major frontend
+// upgrade and we want all users to be sent to a central point.  This function will check for the `forceUpgradeLogoutConfig`
+// configuration map and only run if the last migrated version is lower than the given `migrationVersion`.
+func forceUpgradeLogout(configMapController controllerv1.ConfigMapController, tokenController v3.TokenController, migrationVersion string) error {
+	cm, err := configMapController.Cache().Get(cattleNamespace, forceUpgradeLogoutConfig)
+	if err != nil && !k8serror.IsNotFound(err) {
+		return err
+	}
+
+	// if this is the first ever migration initialize the configmap
+	if cm == nil {
+		cm = &v1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      forceUpgradeLogoutConfig,
+				Namespace: cattleNamespace,
+			},
+			Data: map[string]string{},
+		}
+	}
+
+	// we only want to migrate if the previous installation was earlier than the migration version
+	if lastMigration, ok := cm.Data[rancherVersionKey]; ok {
+		if lastMigration == "dev" || version.Compare(migrationVersion, lastMigration, "<=") {
+			return nil
+		}
+	}
+
+	logrus.Infof("Detected %s upgrade, forcing logout for all web users", migrationVersion)
+
+	// list all tokens that were created for the dashboard
+	allTokens, err := tokenController.Cache().List(labels.SelectorFromSet(labels.Set{tokens.TokenKindLabel: "session"}))
+	if err != nil {
+		logrus.Error("Failed to list tokens for upgrade forced logout")
+		return err
+	}
+
+	// log out all the dashboard users forcing them to be redirected to the login page
+	for i := 0; i < len(allTokens); i++ {
+		err = tokenController.Delete(allTokens[i].ObjectMeta.Name, &metav1.DeleteOptions{})
+		if err != nil && !k8serror.IsNotFound(err) {
+			logrus.Error("Failed to delete token for upgrade forced logout")
+			logrus.Debugf("Failed to delete token [%s] for upgrade forced logout", allTokens[i].Name)
```

Comment:
If the concern is that token's `Name` contains sensitive info, we need to reduce logging to trace here. If that's not the case, we may want to just include the name into Error log message and log once instead of twice.